### PR TITLE
Allow for multiple dependencies/dependents from the same package

### LINKED ISSUE #11983: Incomplete computation of installed dependents
Follow-up of https://github.com/spack/spack/issues/3690#issuecomment-510020811. Basically everything that relies on computing dependents is broken when a spec has more than one dependent with the same name.

### Steps to reproduce the issue
The issue can be reproduced with e.g. Spack at d3be42fccaabbdc71c03da0886214b31a29921e5. Let's start by installing 2 different versions of `hdf5` depending on the same `zlib`:
```console
$ spack install --fake hdf5~mpi+fortran hdf5~mpi~fortran
```
Let's then check that they actually both depend on the same `zlib` and they have different hashes:
```console
$ spack find -ld hdf5
==> 2 installed packages
-- linux-ubuntu18.04-x86_64 / gcc@9.0.1 -------------------------
xyhss36    hdf5@1.10.5
tn4qvs7        ^zlib@1.2.11

4f3j7km    hdf5@1.10.5
tn4qvs7        ^zlib@1.2.11
```
Asking which are the installed dependents of `zlib` will give the wrong answer an **skip one of the two specs**:
```console
$ spack dependents -i zlib
==> Dependents of zlib@1.2.11%gcc@9.0.1/tn4qvs7
-- linux-ubuntu18.04-x86_64 / gcc@9.0.1 -------------------------
xyhss36 hdf5@1.10.5
```
This is ultimately due to how dependents are computed in `spack.spec.Spec`:

https://github.com/spack/spack/blob/c2de2558b6423711e3b2cdbaaca1c0b0dfb61757/lib/spack/spack/spec.py#L1063-L1072

by constructing a map **from the package name** to the branch connecting it to the current spec. When there is more than one dependent with the same name only the last dependent added will be recorded.

This might be fine if we are working on a single tree DAG (where we are ensured there's at most one spec with the same name), but it is not if we are analyzing the entire DAG of all the installed applications. For instance the method:

https://github.com/spack/spack/blob/d3be42fccaabbdc71c03da0886214b31a29921e5/lib/spack/spack/database.py#L878-L911

will return wrong results when we ask to analyze in the `parents` direction.

### Error Message

There's no error message, just the information on dependents of an installed spec is computed wrongly. This lead for instance to issues like #3690 or #5637.

### Information on your system

The platform is not relevant for this bug.

### PR DESCRIPTION
**Pull request by alalazo**
fixes #11983 
closes #16447

This PR changes the internal representation of `Spec` to allow for multiple dependencies or dependents stemming from the same package. This change permits to represent cases which are frequent in cross compiled environments or to bootstrap compilers.

Modifications:
- [x] Substitute `DependencyMap` with `_EdgeMap`. The main differences are that the latter does not support direct item assignment and can be modified only through its API. It also provides a `select_by` method to query items.
- [x] Reworked a few public APIs of `Spec` to get list of dependencies or related edges.
- [x] Added unit tests to prevent regression on #11983 and prove the synthetic construction of specs with multiple deps from the same package. 

~Due to the change in the internal representation of specs, the YAML file for specs will change too and the "dependencies" field will be a list of dictionaries instead of a single dictionary:~
```yaml
    dependencies:
      pkgconf:
      - hash: nip2nwwydp6asi4iiza37drmolecwzyg
        type:
        - build
```
~This in turn will cause all of the hashes to change, so the PR is definitely not backward compatible with old installation hashes.~

Since #22845 went in first, this PR reuses that format and thus it should not change hashes. What happens is that in the list of dependencies we may have the same package being present multiple times with different associated specs.

### CODE REVIEW

**Comment by alalazo** [comment]
@tgamblin Rebased on top of #21618 and ready for review

File: lib/spack/spack/spec.py
Comment (scheibelp) [inline lib/spack/spack/spec.py]: If I understand correctly:

* Representing dependents in this manner fixes https://github.com/spack/spack/issues/11983
* It is not required to change how dependencies are represented to fix https://github.com/spack/spack/issues/11983
* The hash changes because of the change in how dependencies are represented (in other words, changing the representation of dependents does not change the hash)

What use cases do you have in mind when changing the representation of dependencies?

File: lib/spack/spack/spec.py
Comment (scheibelp) [inline lib/spack/spack/spec.py]: Elsewhere it is mentioned that multiple dependencies with the same name was not supported for the old concretizer (and this method is only used for concretization) - so I think it would be appropriate to add a warning similar to the one you added to `_concretize_helper`.

File: lib/spack/spack/spec.py
Comment (scheibelp) [inline lib/spack/spack/spec.py:4653]: I'm guessing the intent would be to resolve this TODO before merging this.

What is the effect of splicing in a dependency when there are multiple instances of the package in the DAG? I would assume that this function might need to specify `deptypes` in that case.

**Review by scheibelp — COMMENTED** [review]
I have a few questions to start with. Primarily I'm interested in why this changes both dependencies and dependents. My understanding is that just changing dependents would be sufficient to fix #11983 (more details in the comments).

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py]: This PR is meant to be part of a more general extension to allow separate concretization of build dependencies. The plan to arrive there is:
1. Refactor specs so that they can model DAGs with the same package present in multiple configurations (this PR)
2. Modify the ASP-based solver to account for cases that it can't currently handle, based on the capabilities introduced in 1.

Examples of 2. can be split dependencies in a cross-compilation environment or things like [the one discussed in this thread](https://spackpm.slack.com/archives/C5W7NKZJT/p1620044403296700?thread_ts=1616663239.174500&cid=C5W7NKZJT). I tried to encode in the unit tests 2 cases that will be of interest for future development, see:

https://github.com/spack/spack/blob/83dcaa2cea448bb42c136d4d22679a42a05f6989/lib/spack/spack/test/spec_dag.py#L1002-L1050

TL;DR This PR solves #11983 incidentally, as a result of generalizing the data structures used to keep track of edges in a DAG. It is not meant as a simple fix for just that issue.

File: lib/spack/spack/spec.py
Comment (scheibelp) [inline lib/spack/spack/spec.py:801]: IMO it would he helpful to clarify that the dictionary maps a package name to a list of edges/`DependencySpec` objects.

File: lib/spack/spack/spec.py
Comment (scheibelp) [inline lib/spack/spack/spec.py]: I'd recommend a 2-value "enum":

```
class EdgeDirection(object):
    parent = 0
    child = 1
```

File: lib/spack/spack/spec.py
Comment (scheibelp) [inline lib/spack/spack/spec.py:858]: This API seems a little strange to me because the edge map is entirely parents or entirely children. I think something like:

```
def select_by(self, name=None, deptypes='all'):
    ...
    if self.store_by == 'child':
        selected = [d for d in selected if d.spec.name == name]
   
```

would avoid that confusion.

File: lib/spack/spack/spec.py
Comment (scheibelp) [inline lib/spack/spack/spec.py:831]: Should this be `is` vs. `==`?

```
if edge.spec is dep.spec and edge.parent is dep.parent:
```

Specs may temporarily compare `==` even if we intend to manage two of them as separate objects (which eventually become unequal).

This may not be critical since the old concretizer should never do something like this and the new concretizer would (I think) only ever use a function like this when the specs were concrete.

File: lib/spack/spack/spec.py
Comment (scheibelp) [inline lib/spack/spack/spec.py]: Because you changed the name (and I agree that a private name is suitable), I don't think you have to worry about handling more references to the function. I think updating the comment like https://github.com/spack/spack/pull/21683#discussion_r625335739 would be sufficient.

**Review by scheibelp — COMMENTED** [review]
I'm still reading through this to understand it. I have some additional questions.

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py:4653]: > I'm guessing the intent would be to resolve this TODO before merging this.

To my understanding splicing is not yet used in Spack and there's just been some commit to prepare for this functionality. We can either leave a TODO and check it in a later PR (since there are no test failures) or see if @becker33 or maybe @nhanford can help with that here.

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py:831]: I think it should be `==` i.e. if we have equivalent nodes defining the edges, we don't care if they are or are not the same object in memory.

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py:858]: I think the API is correct, in the sense that an `_EdgeMap` can store arbitrary list of edges, which are collected in a dictionary by either parent name or child name. What I mean is that it's specs that use `_EdgeMap` objects to store edges having all the same parent or all the same children but an `_EdgeMap` object can in principle store completely different edges in terms of parents and children. Hence I think on this class the API is fine.

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py]: Done in 2d4673b

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py]: Done in 3bfd8fe

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py:801]: Done in 2d4673b

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py]: Done in 2d4673b

**Comment by alalazo** [comment]
@scheibelp Ready for a second review

File: lib/spack/spack/spec.py
Comment (scheibelp) [reply lib/spack/spack/spec.py:831]: If the specs are not concrete, but later differ after becoming concrete, would we lose the edge information (i.e. only record one edge for what ends up later being two distinct dependencies)?

I think `==` is ok as long as we mandate that the specs are concrete.

File: lib/spack/spack/spec.py
Comment (scheibelp) [reply lib/spack/spack/spec.py:858]: > What I mean is that it's specs that use _EdgeMap objects to store edges having all the same parent or all the same children but an _EdgeMap object can in principle store completely different edges in terms of parents and children. 

I'm not sure if by this you mean

* you might want an edge map that stores edges only to parents, but for multiple children
* you might want an edge map that stores edges to both parents and children (I assume not this since the `EdgeMap` has a `store_by_child` property, but I want to make sure)
* something else?

File: lib/spack/spack/spec.py
Comment (scheibelp) [inline lib/spack/spack/spec.py:1500]: If there are multiple parents there may still be a single root (this would occur for a dependency in a single concretized spec), but if there are multiple roots (e.g. after database reconstruction), this can raise an error. I think that would settle it.

File: lib/spack/spack/spec.py
Comment (scheibelp) [reply lib/spack/spack/spec.py:4653]: As is, this TODO is too vague. We should explain why this PR may lead to a failure to respect splice semantics.

For starters, I think this doesn't handle the case where we depend on the same package once for linking and once for building, and we want to splice in a separate build dependency (for that, I think `splice` would need to allow specifying a dependency type). I think that should be handled here (so in that sense it's not a good example of what should go into the TODO), but I also want to consider other cases where the logic is incomplete.

Overall I think this will have to be understood better before this PR is merged (maybe we don't need to solve every problem, but the anticipated issues will need to be enumerated).

(for reference the PR that added splicing is https://github.com/spack/spack/pull/20262)

**Review by scheibelp — CHANGES_REQUESTED** [review]
This includes a couple questions and responses. Largest request is likely that we should think through how this interacts with splice (more in the comments).

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py:858]: > This API seems a little strange to me because the edge map is entirely parents or entirely children.

Think for example storing all the edges for an environment that has multiple root specs and wanting to retrieve all the different edges from e.g. any `hdf5` to any `zlib`. With the current API you can do it with this object.

Another way of saying this is that the current API fit using an EdgeMap object outside of a Spec object as a collection of edges.

**Comment by haampie** [comment]
We will need to extend our spec DSL to support direct vs transient dependencies if we go this way, so that we don't end up with false positives in queries:

Suppose `build-pkg` needs `zstd@:1.4.0` and `my-pkg` requires `build-pkg` and a `zstd` bounded from below: `depends_on('zstd@1.5.0:')`. Then this could concretize s.t. it includes two zstd's at different versions.

But up till now you could equivalently write `depends_on('zstd')` `conflicts('^zstd@:1.4')`. 
After this change it is not, because `^zstd` can match the dependency of `build-pkg` and `my-pkg`, so it will put a constraint on both of them, resulting in concretization failure.

Maybe `conflicts('^^zstd')` should mean: any zstd down the line; and `conflicts('^zstd')`: any direct dependency zstd. But that'd be terribly breaking. Makes me think of `/` and `//` in xpath.

Generally I think 99% of the packages that use `^xyz` in `conflicts` or `when` clauses refer to their direct dependencies, not to some arbitrary dependency of a dependency of a dependency.

**Comment by alalazo** [comment]
> Generally I think 99% of the packages that use ^xyz in conflicts or when clauses refer to their direct dependencies, not to some arbitrary dependency of a dependency of a dependency.

I think that's already the semantics for packages. If you conflict with something it's supposed to be either a node attribute or a direct dependency (maybe through a virtual). We'll need to extend the DSL used on the command line in #15569 to allow referencing edge attributes, of which virtuals are an example.

Anyhow, this PR just changes the internal data representation, so no user facing behavior is changed. There are APIs to  discriminate among transitive dependencies and the implementation for `__getitem__` privileges link/run dependencies to build deps.

In a followup PR I'll build on this one to introduce separate concretization of build dependencies, and there we may have to face cases like the one you mentioned above.

File: lib/spack/spack/solver/asp.py
Comment (becker33) [inline lib/spack/spack/solver/asp.py:1884]: This will be slightly tricky to update with separate concretization of build deps, but I think the PSIDs will give us a way to make it work.

File: lib/spack/spack/installer.py
Comment (becker33) [inline lib/spack/spack/installer.py:232]: I'm going to be so happy to remove this kludge when we get compilers as deps working...

File: lib/spack/spack/database.py
Comment (becker33) [inline lib/spack/spack/database.py]: Why separate these lines?

File: lib/spack/spack/graph.py
Comment (becker33) [inline lib/spack/spack/graph.py]:
```suggestion
    def children(specs):
        return list(set(itertools.chain.from_iterable(
            s.dependencies(deptype=deptype) for s in specs
        )))

    def parents(specs):
        candidates = list(set(itertools.chain.from_iterable(
            s.dependents(deptype=deptype) for s in specs
        )))
        return [x for x in candidates if x.name in nodes]
```

File: lib/spack/spack/graph.py
Comment (becker33) [inline lib/spack/spack/graph.py]: I think this still assumes there's only one node of a given name in the tree we're sorting -- I don't think this can graph the synthetic creations you use for testing.

File: lib/spack/spack/graph.py
Comment (becker33) [inline lib/spack/spack/graph.py]: If there are multiples of a node, we should use the `diff` capability to create the minimal specs to differentiate them.

File: lib/spack/spack/graph.py
Comment (becker33) [inline lib/spack/spack/graph.py]: Rename the variable to `nodes` if there may be multiple

File: lib/spack/spack/test/database.py
Comment (becker33) [inline lib/spack/spack/test/database.py:943]: What does the number of incoming nodes of a single name mean in this context? Is that just that it's an N-type dependency?

File: lib/spack/spack/test/database.py
Comment (becker33) [inline lib/spack/spack/test/database.py:954]: Would be easier to read in the other order, but not worth changing unless you're in this file anyway

File: lib/spack/spack/spec.py
Comment (becker33) [reply lib/spack/spack/spec.py:831]: If we turn it into one edge, then we know they will concretize the same way because the concretizer will treat them as a single spec.

File: lib/spack/spack/spec.py
Comment (becker33) [inline lib/spack/spack/spec.py]:
```suggestion
        """Trim the dependencies and dependents of this spec."""
```

File: lib/spack/spack/spec.py
Comment (becker33) [inline lib/spack/spack/spec.py:1303]: Could this be replaced with

```
deps = self.dependencies(name=name)
assert len(deps) == 1
return deps[0]
```

**Review by becker33 — CHANGES_REQUESTED** [review]

**Comment by becker33** [comment]
This also needs to be updated for splicing

File: lib/spack/spack/database.py
Comment (alalazo) [reply lib/spack/spack/database.py]: No particular reason, I possibly did it to stop at line 1529 during debugging. Fixed in 8df0af7e8a14f387962f61c530298ea90a84613d

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py]: Fixed in eaeb40c70aee1900c7c93887471b5ce43c1251d5

File: lib/spack/spack/spec.py
Comment (alalazo) [reply lib/spack/spack/spec.py:1303]: Almost. This should be fixed in b4710a36205f386de20320c57a5ce9ae2a751201

File: lib/spack/spack/test/database.py
Comment (alalazo) [reply lib/spack/spack/test/database.py:943]: No, it's really about how many parent specs this spec has. Our mock database looks like this:

https://github.com/spack/spack/blob/b7f4a421e4028115ba437fab7080c30e41d1b0d1/lib/spack/spack/test/conftest.py#L680-L699

so `dyninst` has 3 `callpath` parents. This just checks that we can retrieve all the three edges. Before this PR it was not possible since we couldn't distinguish among different `callpath` specs in `Spec._detendents`.

To be clear, the dependency type is just a edge property - so a `build,link` dependency from spec `foo/abcd` to spec `bar/1234` still counts a a single edge.

**Comment by alalazo** [comment]
@becker33 With respect to what we discussed in Slack to extend the `splice` API to:
```python
def splice(self, other, transitive, deptype):
     ...
```
where `deptype` is used to select which spec we want to splice in `self`, I think the API might not be general enough to cover for cases that may happen when allowing multiple specs from the same package.

One (synthetic) use case that we can't cover with that interface is the following. Start with a spec like:
```
   A
 /   \
B     C
``` 
and  say that we want to splice in that:
```
     C'
   /  \
  /    \
B_1'     B_2'
```
where the new `C` spec depends on 2 different `B` specs. In that case it's not a matter of selecting which node in the original spec needs to be spliced, but rather which one of `B_1'` or `B_2'`needs to be used for the spliced spec if `transitive=True`.[^1] I'm trying to work this out on paper, but I think the might need some "translation" dict that maps hashes of specs that are to be substituted to hashes of specs that will substitute them. We might discuss this again offline, but I wanted to leave here a sketch of the use case.

@nhanford You might be interested in this case / have opinions on how to solve this issue.

[^1]: Another sensible option is to use all three `B`, `B_1'` and `B_2'` in the spliced spec.

File: lib/spack/spack/graph.py
Comment (alalazo) [reply lib/spack/spack/graph.py]: I refactored this function to return specs instead of spec names, and use either `id(spec)` or `spec.full_hash()` to uniquely identify specs depending on the context.

File: lib/spack/spack/graph.py
Comment (alalazo) [reply lib/spack/spack/graph.py]: With the latest code, using this script:
```python
import spack.graph
import spack.hash_types
import spack.spec

spack.hash_types.package_hash = spack.hash_types.SpecHashDescriptor(
    deptype=(), package_hash=True, name='package_hash', override=lambda s: ''
)

Spec = spack.spec.Spec

nodes = {
    'n1': Spec('n1@1.0.0'), 'n2': Spec('n2@1.0.0'), 'n3': Spec('n2@2.0.0'), 'n4': Spec('n3@1.0.0')
}

for s in nodes.values():
    s._mark_concrete()

original = nodes['n1']
nodes['n2'].add_dependency_edge(nodes['n4'], deptype=('build', 'link'))
nodes['n3'].add_dependency_edge(nodes['n4'], deptype=('build', 'link'))
original.add_dependency_edge(nodes['n2'], deptype='run')
original.add_dependency_edge(nodes['n3'], deptype='link')

spack.graph.graph_ascii(original)
```
I obtain the following graph:
```
o n1@1.0.0/bysqrwy
|\
| o n2@2.0.0/tnvdtkr
o | n2@1.0.0/n4gla5x
|/
o n3@1.0.0/gaeiwev
```

File: lib/spack/spack/graph.py
Comment (alalazo) [reply lib/spack/spack/graph.py]: Outdated

**Comment by alalazo** [comment]
This is kind of weird. I force pushed the branch to update the PR. The PR wasn't updated and now I cannot reopen it.

**Pull request closed without merging**
